[DNM] net/pkt: Optimize length computation on a net_pkt

tbursztyka opened the pull request:
I had this for quite a while. In fact, when reworking the net_pkt API, at first, I added this attribute. But it was growing the net_pkt, and was out of scope of the rework anyway.

Sending it to see CI's result.

Commit:
net/pkt: Optimize length computation on a net_pkt

Currently it goes through the chain of net_buf composing the buffer. But
if we have an attribute to store the total, and always keeps it updated
after each operations on the buffer (adding or removing data),  it will
be faster: to need to compute it anymore.

Downside: it grows net_pkt by 2 bytes.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/udp: Check UDP header's length attribute

Length should be at least of UDP header size but not bigger than actual
payload size.

Reported-by: Ruslan Mstoi <ruslan.mstoi@intel.com>

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
tests/net: Fix IPv6 and UDP test as UDP got stricter on checking header

Now length of the UDP header is checked, and obviously these test were
avoiding setting it properly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/udp: Always drop IPv6 UDP packet if checksum field is 0

Missing checksum is only valid in IPv4 and only on certain context.

Fixes #16483

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jukkar commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1735,28 +1737,36 @@ size_t net_pkt_remaining_data(struct net_pkt *pkt)
+		return -EINVAL;
+	}
+
+	pkt->length -= length;
```
Doing this without any checks looks dangerous. What if pkt->length goes <0 and becomes a large value.

tbursztyka replied:
It is unlikely to happen if, and only if, all operations are done through net_pkt API. But we cannot prevent users to mess up with net_buf I guess. So let's be conservative then yes

jukkar commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1735,28 +1737,36 @@ size_t net_pkt_remaining_data(struct net_pkt *pkt)
 
 int net_pkt_update_length(struct net_pkt *pkt, size_t length)
 {
+	size_t len = length;
```
I think we should keep len as a signed value, then check below that it is >0

jukkar commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1735,28 +1737,36 @@ size_t net_pkt_remaining_data(struct net_pkt *pkt)
 {
 	struct net_pkt_cursor *c_op = &pkt->cursor;
 	struct net_pkt_cursor backup;
+	size_t len = length;
```
same here

tbursztyka commented:
I missed the fact that L2 use net_buf_pull() 

And net_pkt_pull isn't the same. I'll rename it to net_pkt_shrink() (or something more semantically relevant) and a new net_pkt_pull() will do the same as net_buf_pull(), always on the first buf (and will reinitialize the cursor) so length attribute will be updated accordingly

Commit:
net/pkt: Rename net_pkt_pull() into net_pkt_remove()

It is semantically more relevant that way.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/pkt: Add a net_pkt_pull() function

This will be used, for instance, in L2 where net_buf_pull() is being
used but without updating net_pkt's length which is bogus.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/tests: Adapt tests so net_pkt's length attribute is used relevantly

There were places where net_pkt was not filled in via net_pkt's API.
(Using legacy direct memory access etc...) Such technics were not
updating net_pkt's length then.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
the first 2 commits comes from #16471 so no need to review those

tbursztyka commented:
and the third come from #16484, same don't review that one

jukkar commented on include/net/net_pkt.h:
```diff
@@ -701,7 +703,7 @@ static inline void net_pkt_set_timestamp(struct net_pkt *pkt,
 
 static inline size_t net_pkt_get_len(struct net_pkt *pkt)
 {
-	return net_buf_frags_len(pkt->frags);
```
BTW, should we have a way to recompute the length. So something like
```
size_t net_pkt_recompute_len(pkt)
{
   pkt->length = net-buf_frags_len(pkt->frags);
   return pkt->length;
}
```
This would be handy if we want to make sure the net_pkt contains a valid length. For example if we have a buffer chain and want to attach that to net_pkt, I do that in PPP actually.

tbursztyka replied:
Then use net_pkt_append_buffer(), it will append the buf chain and update the length as well (instead of recomputing it from the start it will just add the length of the chain)

jukkar commented on include/net/net_pkt.h:
```diff
@@ -1589,7 +1589,7 @@ size_t net_pkt_remaining_data(struct net_pkt *pkt);
  * @brief Update the overall length of a packet
  *
- * @details Unlike net_pkt_pull() below, this does not take packet cursor
+ * @details Unlike net_pkt_remove() below, this does not take packet cursor
```
I think your previous suggested name net_pkt_shrink() was better. The remove word sounds like we are removing the net_pkt.

tbursztyka replied:
ok

jukkar commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1797,6 +1797,16 @@ int net_pkt_remove(struct net_pkt *pkt, size_t length)
+int net_pkt_pull(struct net_pkt *pkt, size_t length)
+{
+	if (pkt->buffer) {
+		net_buf_pull(pkt->buffer, length);
```
We should check here that if the net_buf->len is smaller than what we try to pull, the first net_buf should be removed in that case (and the second one if the pull length is large enough etc)

tbursztyka replied:
indeed, without this it's a memory leak issue ahead of us.
